### PR TITLE
Is ld 12 sort shopping list

### LINKED ISSUE #12: 12. As a user, I want to view a list of my shopping list items in order of how soon I am likely to need to buy each of them again so that it’s clear what I need to buy soon.
## Summary
Sorting items by how soon the user might need to buy them will make it easier for the user to identify what they need to buy, which will in turn help them better plan their shopping trips. To that end, we will be sorting items into 4 possible groups, according to how urgently the user should buy the item:

1. Need to buy soon (7 days or fewer until the next purchase)
2. Need to buy kind of soon (between 7 & 30 days until the next purchase)
3. Need to buy not soon (30 days or more until the next purchase)
4. Inactive (60 days have passed since the last purchase)

To implement this feature, your team will need to create some new UI elements so the user knows how their items are sorted and write some logic behind-the-scenes to perform the sorting itself. Read the acceptance criteria and our notes carefully to be sure you understand the tasks at hand, and don’t be afraid to reach out to the rest of your team, and your mentors if you need help. You’ve got this! 

## Acceptance criteria
- [ ] Items in the list are shown with an indicator that tells the user they should buy the item “soon”, “kind of soon”, or “not soon”; or that the item is “inactive”
	- [ ] This urgency indicator *does not* rely on only color
- [ ] `api/firestore.js` exports a new `comparePurchaseUrgency` function with the following behaviors
	- [ ] sorts inactive items last, then
	- [ ] sorts items in ascending order of days until purchase, and
	- [ ] sorts items with the same days until purchase alphabetically

### A stretch goal
If you complete all of the previous acceptance criteria, consider what happens when an item’s `dateNextPurchased` has passed, but it isn’t yet inactive. Let’s call that item “overdue”.

- [ ] Extend the functionality of `comparePurchaseUrgency` to sort “overdue” items to the top of the list
- [ ] Indicate in your UI when an item is overdue

## Notes
There's a lot to do for this issue! Read this section for some ideas about how to approach the tasks ahead. The `getDaysBetweenDates` function in `src/utils/dates.js` will be useful!

Remember: The dates in the Firestore database are special `Timestamp` objects. You’ll need to call their `toDate()` method to get a JavaScript Date that we can work with.

### Indicating purchase urgency in the UI
It’s tempting to use colors or icons to indicate how soon the user should buy an item. Keep in mind:
- If you use only color, a colorblind user won’t be able to tell the difference between urgency indicators
- If you use icons, you have to teach users what those icons mean

With these constraints, it’s probably best to render plain text, like “soon”, “kind of soon”, etc. in order to show urgency. You can still use color alongside that text, and the design of where that text goes is up to you!

### Sorting the list
We’re sorting by multiple criteria:
1. Whether the item is active or inactive, then
2. By the number of days until its `dateNextPurchased`, and
3. Alphabetically by name, if they have the same number of days until purchase

Sorting by multiple criteria can be tricky! Refer to the MDN docs for [`Array.prototype.sort`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) and refresh yourself on how to write *compare functions* so you can write the `comparePurchaseUrgency` function.

### PR DESCRIPTION
**Pull request by isaiahasanchez**
## Description

Added comparePurchaseUrgency function that can be used to sort items by the urgency of which the user is expected to need to buy the item again. Used a sort function with comparePurchaseUrgency on all the data in useShoppingListData. This ensured all data would appear in the proper sorting order. 

Added text indicators on the list items. Added styling as well based on the color preference of a dark background. Chose to add styling via listItem.css for code simplicity. Indicators that were added were "soon", "kind of soon", "not soon", "inactive" and we chose to also include "overdue" for items for which the dateNextPurchased had passed.

## Related Issue

closes #12 

## Acceptance Criteria 

- [x] Items in the list are shown with an indicator that tells the user they should buy the item “soon”, “kind of soon”, or “not soon”; or that the item is “inactive”
	- [x] This urgency indicator *does not* rely on only color
- [x] `api/firestore.js` exports a new `comparePurchaseUrgency` function with the following behaviors
	- [x] sorts inactive items last, then
	- [x] sorts items in ascending order of days until purchase, and
	- [x] sorts items with the same days until purchase alphabetically
- [x] Extend the functionality of `comparePurchaseUrgency` to sort “overdue” items to the top of the list
- [x] Indicate in your UI when an item is overdue

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before
Screenshot 2023-09-19 at 12 38 56 PM

### After
Screenshot 2023-09-19 at 12 43 44 PM

## Testing Steps / QA Criteria

inactive items: have a dateLastPurchased more than 60 days in the past
overdue items: an item with dateNextPurchased in the past 
soon items: an item with dateNextPurchased within 7 days or less from now
kind of soon items: an item with dateNextPurchased between 7 and 30 days from now
not soon: an item with dateNextPurchased over 30 days from now

- open in GH preview 
- navigate to desired list 
- add one or more examples of each type of item as defined above (make changes in firebase console to the item's dates in order to set item to a certain indicator category)
- verify that the items are sorted based on the categories first, then by date and if the dates are the same then alphabetically
- verify that the indicator appears clearly and correctly

### CODE REVIEW

File: src/components/ListItem.jsx [hunk 1]
```diff
@@ -19,6 +20,30 @@ export function ListItem({ item, listId }) {
...
+			return 'overdue';
+		} else if (days <= 7) {
+			return 'soon';
+		} else if (days > 7 && days <= 30) {
+			return 'kind of soon';
+		} else {
+			return 'not soon';
+		}
+	}
+	const indicatorClass = `indicator-${determineItemIndicator(item).replaceAll(
```
Comment (patty-rose) [inline src/components/ListItem.jsx:42]: Just a curiosity/best practice comment, no need to revise. I wonder if it's more React best practice to declare this state to be at the top of the component and within a useState hook. If done this way we might need to use a useEffect hook to trigger setting the state when the component mounts......... really not sure if this is totally off.

**Review by patty-rose — APPROVED** [review]
Wow 🥇 I am so impressed with how quickly you all pulled this off. I felt confused reading the issue and acceptance criteria but then when I followed your testing steps and read through your code it made total sense! Great work!

File: src/components/ListItem.jsx (hunk 1 above)
Comment (isaiahasanchez) [reply src/components/ListItem.jsx:42]: are you asking about indicatorClass? It isn't state it is just a variable to use to get the css class for the indicator? I believe if I understand your questions correctly?

Comment (patty-rose) [reply src/components/ListItem.jsx:42]: I was asking about indicatorClass. Oh yeah I guess I'm confused about what is considered state, so I think you're right that it's not state.

**Review by shannonerhea — APPROVED** [review]
Testing steps are easy to follow. Meets AC. Great job Team!!

File: src/components/ListItem.css [hunk 2]
```diff
@@ -12,3 +13,23 @@
 .ListItem-label {
 	margin-left: 0.2em;
 }
+
+.indicator-soon {
+	color: orange;
+}
+
+.indicator-kind-of-soon {
+	color: yellow;
```
Comment (MikeBLambert) [inline src/components/ListItem.css]: I think that the yellow may not have enough color contrast from an accessibility standpoint.

Here is a screenshot when analyzing with Wave 

screenshot

File: src/components/ListItem.jsx [hunk 3]
```diff
@@ -29,6 +54,8 @@ export function ListItem({ item, listId }) {
 				/>
 				{name}
 			</label>
+			<br />
+			<p className={indicatorClass}>{determineItemIndicator(item)}</p>
```
Comment (MikeBLambert) [inline src/components/ListItem.jsx]: Should this be part of the label? Maybe a `span` inside of the label?

File: src/components/ListItem.jsx (hunk 1 above)
Comment (MikeBLambert) [reply src/components/ListItem.jsx:42]: Great question @patty-rose! I think that this is better off as a variable b/c if we were to declare as state and then update that state every time that `item` updates, then we'd be causing an unnecessary re-render. Also, listening for `item` updates via the dependency array gets trigger b/c as an object it would technically retrigger the dependency array on every render unless we did some fancy stuff.

Long story short, I usually try to minimize declaring things as state unless it's an operation that requires a bunch of processing power and want to run that function as little as possible rather than on every re-render. In this case it's great as is. :-)

**Review by MikeBLambert — CHANGES_REQUESTED** [review]
This is looking awesome! I'd like us to figure out those accessibility things that I pointed out before we consider this one done. Great work so far!

File: src/components/ListItem.jsx (hunk 3 above)
Comment (lydiadunning) [reply src/components/ListItem.jsx]: I think making it part of the label would be best, since clicking on the label will check the checkbox.   It all shows up on the same line as the checkbox because of CSS.

File: src/components/ListItem.css (hunk 2 above)
Comment (lydiadunning) [reply src/components/ListItem.css]: We decided to choose our colors based on how the site looks with a dark background, which seemed to be the default based on the current CSS. We can add an alternative to the colors for light backgrounds with a media query.

**Comment by lydiadunning** [comment]
Updated with changes. Moved the text into a span in the label, and used css variables set in ListItem and in a media query to use different colors in light and dark mode.
Also removed formatting styling from ListItem, where it didn't have an impact, and added grid styling to the label. This looks neater than flexbox, but may complicate adding the delete button to the ListItem.

**Review by MikeBLambert — APPROVED** [review]
Looks great!
